drivers/dcf77/dcf77.c

```diff
@@ -0,0 +1,203 @@
+#define DCF77_READING_CYCLE                     (59)
+
+#define DCF77_MINUTE_MASK                       (0xFE00000ULL)
+#define DCF77_HOUR_MASK                         (0x7E0000000ULL)
+#define DCF77_DATE_MASK                         (0x1FFFFFC00000000ULL)
+
+#define DCF77_MINUTE_SHIFT                      (21)
+#define DCF77_HOUR_SHIFT                        (29)
+#define DCF77_DATE_SHIFT                        (36)
+
+static void _level_cb_high(dcf77_t *dev)
+{
+    switch (dev->internal_state) {
+        case DCF77_STATE_START:
+            DEBUG("[dcf77] EVENT START 1 !\n");
+            dev->stopTime = xtimer_now_usec();
+            if ((dev->stopTime - dev->startTime) >
+                DCF77_PULSE_START_HIGH_THRESHOLD) {
+                memset(&dev->bitseq.bits, 0, sizeof dev->bitseq.bits);
+                dev->internal_state = DCF77_STATE_RX;
+            }else{
+                dev->internal_state = DCF77_STATE_IDLE;
+            }
+            break;
+        case DCF77_STATE_RX:
+            DEBUG("[dcf77] EVENT RX 1 !\n");
+            dev->startTime = xtimer_now_usec();
+            break;
+    }
+}
+
+static void _level_cb_low(dcf77_t *dev)
+{
+    switch (dev->internal_state) {
+        case DCF77_STATE_IDLE:
+            DEBUG("[dcf77] EVENT IDLE 0  !\n");
+            dev->startTime = xtimer_now_usec();
+            dev->internal_state = DCF77_STATE_START;
+            break;
+        case DCF77_STATE_RX:
+            DEBUG("[dcf77] EVENT RX 0 !\n");
+            dev->stopTime = xtimer_now_usec();
+            if ((dev->stopTime - dev->startTime) >
+                DCF77_PULSE_WIDTH_THRESHOLD) {
+                dev->bitseq.bits |=  1ULL << dev->bitCounter;
+            }
+            dev->bitCounter++;
+            if (dev->bitCounter >= DCF77_READING_CYCLE) {
+                dev->bitCounter = 0;
+                dev->startTime = xtimer_now_usec();
+                dev->last_bitseq.bits = dev->bitseq.bits;
+                dev->internal_state = DCF77_STATE_START;
+            }
+            break;
+    }
+}
+
+static void _level_cb(void *arg)
+{
+    dcf77_t *dev = (dcf77_t *)arg;
+
+    if (gpio_read(dev->params.pin)) {
+        _level_cb_high(dev);
+    }
+    else {
+        _level_cb_low(dev);
+    }
+}
+
+/**
+ * @brief   Initialize the Device
+ *
+ * @param   dev     device
+ * @param   params  device_params
+ *
+ * @retval  DCF77_OK             Success
+ * @retval  DCF77_INIT_ERROR     Error in initialization
+ */
+
+int dcf77_init(dcf77_t *dev, const dcf77_params_t *params)
+{
+    DEBUG("dcf77_init\n");
+
+    /* check parameters and configuration */
+    assert(dev && params);
+    dev->params = *params;
+    dev->internal_state = DCF77_STATE_IDLE;
+    dev->bitCounter = 0;
+    if (!gpio_init_int(dev->params.pin, dev->params.in_mode, GPIO_BOTH,
+                       _level_cb,
+                       dev)) {
+        return DCF77_OK;
+    }
+    else {
+        gpio_irq_disable(dev->params.pin);
```

Comment:
If initializing the GPIO failed, you don't have to disable it - that would probably fail too.